hoomd/snapshot.py

```diff
@@ -109,3 +110,104 @@ def replicate(self, nx, ny, nz):
 
     def _broadcast_box(self):
         self._cpp_obj._broadcast_box(self._comm.cpp_mpi_conf)
+
+    @classmethod
+    def _from_gsd_snapshot(cls, gsd_snap, communicator):
+        gsd_snap.validate()
+        if communicator.rank == 0:
+            snap = cls(communicator=communicator)
+
+            # Set all particle attributes in snap from gsd_snap
+            if gsd_snap.particles.N:
+                snap.particles.N = gsd_snap.particles.N
+            if gsd_snap.particles.types:
+                snap.particles.types = gsd_snap.particles.types
+            if gsd_snap.particles.angmom is not None:
+                snap.particles.angmom[:] = gsd_snap.particles.angmom
+            if gsd_snap.particles.body is not None:
+                snap.particles.body[:] = gsd_snap.particles.body
+            if gsd_snap.particles.charge is not None:
+                snap.particles.charge[:] = gsd_snap.particles.charge
+            if gsd_snap.particles.diameter is not None:
+                snap.particles.diameter[:] = gsd_snap.particles.diameter
+            if gsd_snap.particles.image is not None:
+                snap.particles.image[:] = gsd_snap.particles.image
+            if gsd_snap.particles.mass is not None:
+                snap.particles.mass[:] = gsd_snap.particles.mass
+            if gsd_snap.particles.moment_inertia is not None:
+                snap.particles.moment_inertia[:] = gsd_snap.particles.moment_inertia
+            if gsd_snap.particles.orientation is not None:
+                snap.particles.orientation[:] = gsd_snap.particles.orientation
+            if gsd_snap.particles.position is not None:
+                snap.particles.position[:] = gsd_snap.particles.position
+            if gsd_snap.particles.typeid is not None:
+                snap.particles.typeid[:] = gsd_snap.particles.typeid
+            if gsd_snap.particles.velocity is not None:
+                snap.particles.velocity[:] = gsd_snap.particles.velocity
```

Comment:
For these it would be more concise to do something like

```python

for property_ in ('positions', 'angmom', 'body', 'charge', 'diameter', ...):
    gsd_prop = getattr(gsd_snap.particles, property_)
    if gsd_prop is not None:
        setattr(snap.particles, gsd_prop)
```

I don't care strongly either way but I do prefer the more concise style. If you decide to implement this, something similar can be done for bonds, angles, etc.